feat(auth): implement JWT authentication system

MaksymChukhrai opened the pull request:
## Description of changes
- Implemented complete JWT authentication system with access and refresh tokens
- Added JWT Guards for route protection with @Public() decorator
- Configured password hashing with bcrypt (10 rounds)
- Created token refresh mechanism with rotation
- Added Swagger documentation for all endpoints
- All 6 auth endpoints tested and verified working

## Issue ticket code (and/or) and link
- Task: **JWT Authentication System**
- Related: `auth-system` branch - **potential merge conflicts**

## 🏗️ Architecture Overview

### JWT Strategies
- **JwtStrategy** - validates access tokens (15 min expiration)
- **JwtRefreshStrategy** - validates refresh tokens (7 days expiration)

### Guards
- **JwtAuthGuard** - global guard (APP_GUARD provider)
- Bypass with `@Public()` decorator for public routes

### Decorators
- **@Public()** - marks endpoints as public (no JWT required)
- **@CurrentUser()** - injects authenticated user from JWT payload

### Endpoints Implemented
1. ✅ `POST /auth/register` - user registration + JWT tokens
2. ✅ `POST /auth/login` - user login + JWT tokens
3. ✅ `POST /auth/refresh` - refresh tokens (public)
4. ✅ `POST /auth/logout` - user logout (protected)
5. ✅ `GET /auth/profile` - current user data (protected)
6. ✅ `GET /auth/all` - all users list (public demo endpoint)

## 🧪 Testing
- ✅ All endpoints tested via Swagger UI at `/api`
- ✅ JWT token validation working correctly
- ✅ Refresh token rotation working
- ✅ Guards protecting routes as expected
- ✅ @Public() decorator bypassing guards properly
- ✅ Password hashing verified
- ✅ Token expiration handling tested

## 📦 Dependencies Added
```json
{
  "@nestjs/jwt": "^11.0.0",
  "@nestjs/passport": "^11.0.5",
  "passport-jwt": "^4.0.1",
  "bcrypt": "^5.1.1",
  "@types/bcrypt": "^5.0.2",
  "@types/passport-jwt": "^4.0.1"
}
```
⚠️ **IMPORTANT: Merge Conflict Warning**

**Related branch:** `auth-system`  
Both branches modify authentication logic:  
- **This PR:** Full JWT authentication system (foundation)  
- **auth-system branch:** Magic link email authentication  

🎯 **Suggested Merge Strategy:**  
**Recommended approach:**  
✅ Merge this PR first (JWT foundation)  
⏭️ Magic link builds on top of JWT infrastructure:  
- Generate JWT token in magic link email  
- Use existing `JwtStrategy` for validation  
- Reuse `@Public()` decorator  
- Leverage existing guards  

**Benefits:**  
- No code duplication  
- Consistent authentication flow  
- Magic link becomes a JWT token delivery method  
- Reduced merge conflicts  

**Alternative:**  
Merge both independently → manual conflict resolution required  

---

🔐 **Security Features**  
✅ Passwords hashed with `bcrypt` (10 salt rounds)  
✅ JWT secrets from environment variables  
✅ Refresh tokens stored in database (can be revoked)  
✅ Access tokens short-lived (15 minutes)  
✅ Refresh tokens long-lived (7 days)  
✅ Token rotation on refresh (old token invalidated)  

🚀 **Production Improvements (Future):**  
- Hash refresh tokens before storing in DB  
- Implement token blacklist for immediate logout  
- Add rate limiting on auth endpoints  
- Enable CORS with proper origins  
- Add refresh token rotation detection  

---

📸 **Testing Evidence**  
All endpoints tested via **Swagger UI**:  
✅ `POST /auth/register` → 201 Created  
✅ `POST /auth/login` → 200 OK  
✅ `POST /auth/refresh` → 200 OK (new tokens)  
✅ `POST /auth/logout` → 200 OK  
✅ `GET /auth/profile` → 200 OK (protected)  
✅ `GET /auth/all` → 200 OK (public)  

**Test scenarios covered:**  
- Register → Login → Profile access  
- Token refresh flow  
- Logout invalidates refresh token  
- Protected routes require JWT  
- Public routes work without JWT

🤝 **Ready for Review**  
Luda, Please review and advise on merge strategy with `auth-system` branch.  
All code follows team conventions and passes TypeScript/ESLint checks.  
✅ Unit and integration tests completed with no failures  
✅ Swagger UI verification for all authentication endpoints  
📌 Next step: Decide merge order to minimize conflicts and align magic link implementation with JWT foundation

MaksymChukhrai committed:
feat(auth): implement JWT authentication system

- Add JWT strategies (access + refresh tokens)
- Implement JWT guards for route protection
- Create @Public() and @CurrentUser() decorators
- Add password hashing with bcrypt
- Implement token refresh mechanism
- Add Swagger documentation for all auth endpoints
- Configure global JWT guard with public route exceptions

Endpoints implemented:
- POST /auth/register - user registration
- POST /auth/login - user login
- POST /auth/refresh - token refresh
- POST /auth/logout - user logout
- GET /auth/profile - current user (protected)
- GET /auth/all - all users (public)

Tech stack:
- @nestjs/jwt ^11.0.0
- @nestjs/passport ^11.0.5
- passport-jwt ^4.0.1
- bcrypt for password hashing
- TypeORM + MySQL

All endpoints tested via Swagger and working correctly.

lyudad commented on src/main.ts:
```diff
@@ -1,20 +1,38 @@
 }
-bootstrap();
+
+// Исправили floating promise
```
only english text. and this comment is not needed

lyudad commented on src/modules/auth/auth.module.ts:
```diff
@@ -1,12 +1,41 @@
+    JwtRefreshStrategy,
+    {
+      provide: APP_GUARD,
+      useClass: JwtAuthGuard, // Global JWT guard
```
its always global

lyudad commented on src/modules/auth/auth.service.ts:
```diff
@@ -16,34 +28,111 @@ export class AuthService {
+        throw new UnauthorizedException('Invalid refresh token');
+      }
+
+      // TODO: Add hash comparison to production
```
why to production?

lyudad commented on src/modules/auth/user.entity.ts:
```diff
@@ -1,13 +1,27 @@
   username: string;
 
   @Column()
+  @Exclude() // Do not return in API responses
```
Why  comment needed?

lyudad commented on src/modules/auth/auth.controller.ts:
```diff
@@ -11,25 +11,97 @@ import {
+  @Post('logout')
+  @HttpCode(HttpStatus.OK)
+  @ApiOperation({ summary: 'Logout user' })
+  public async logout(@CurrentUser() user: UserEntity) {
```
1.1.1

MaksymChukhrai committed:
fix(auth): resolve code quality issues from team review

- Remove unnecessary comments (main.ts, auth.module.ts, user.entity.ts)
- Add explicit return types to all controller methods (rule 1.1.1)
- Add JwtPayload interface for type-safe JWT handling
- Remove TODO comments as per rule 1.1.10
- Clean up redundant inline documentation

All code now complies with team code quality requirements.

MaksymChukhrai committed:
fix(auth): add missing JwtPayload interface file

- Add jwt-payload.interface.ts for type-safe JWT token handling
- Required by auth.service.ts for token verification

lyudad commented on src/modules/auth/auth.service.ts:
```diff
@@ -1,55 +1,232 @@
 import { UserResponseDto } from './dto/user-response.dto';
+import { ConfigService } from '@nestjs/config';
+
+import { JwtPayload } from '../../types/jwt-payload.interface';
```
project have alias, please use it

lyudad commented on src/modules/auth/auth.service.ts:
```diff
@@ -1,55 +1,232 @@
+   * @returns User profile data
+   * @throws UnauthorizedException if user not found
+   */
+  public async getProfile(userId: string): Promise<UserResponseDto> {
```
but we will not have a profile. We only need to gather the email of the users who use service

lyudad commented on src/modules/auth/auth.service.ts:
```diff
@@ -1,55 +1,232 @@
+   * Demo endpoint - should be removed in production
+   * @returns List of all users
+   */
+  public async getAllUsers(): Promise<UserResponseDto[]> {
```
same, not needed
